### PR TITLE
EKF RingBuffer avoid unnecessary copying

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

File: EKF/control.cpp [hunk 1]
```diff
@@ -197,12 +199,8 @@ void Ekf::controlExternalVisionFusion()
...
 
 				// add the reset amount to the output observer buffered data
-				outputSample output_states;
-				unsigned output_length = _output_buffer.get_length();
-				for (unsigned i=0; i < output_length; i++) {
-					output_states = _output_buffer.get_from_index(i);
-					output_states.quat_nominal = _state_reset_status.quat_change * output_states.quat_nominal;
-					_output_buffer.push_to_index(i, output_states);
+				for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+					_output_buffer[i].quat_nominal *= _state_reset_status.quat_change;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply EKF/control.cpp]: Testing using a ground based magnetic anomaly to force a yaw reset shows that the order has not been changed so it appears that q1 *= q2 is the same as q1 = q2 * q1;

File: EKF/ekf_helper.cpp [hunk 2]
```diff
@@ -306,48 +295,34 @@ void Ekf::resetHeight()
...
-
-	for (unsigned i = 0; i < output_length; i++) {
-		output_states = _output_buffer.get_from_index(i);
-		output_states.quat_nominal = q_delta * output_states.quat_nominal;
-		output_states.quat_nominal.normalize();
-		output_states.vel += vel_delta;
-		output_states.pos += pos_delta;
-		_output_buffer.push_to_index(i, output_states);
+	for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+		_output_buffer[i].quat_nominal *= q_delta;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply EKF/ekf_helper.cpp]: This has been resolved. q1 *= q2 is the same as q1 = q2*q1

**Comment by dagar** [comment]
I glossed over the ordering. Probably better to make it explicit.

File: EKF/RingBuffer.h [hunk 3]
```diff
@@ -45,54 +45,51 @@ template <typename data_type>
...
 			_buffer[index].time_us = 0;
 		}
 		_first_write = true;
+
 		return true;
 	}
 
-	void unallocate()
-	{
-		if (_buffer != NULL) {
```
[the remark this answers is not in the document]
Comment (priseborough) [reply EKF/RingBuffer.h]: This is OK with c++

[the remark this answers is not in the document]
Comment (dagar) [reply EKF/RingBuffer.h]: In c++ you can delete NULL (it checks internally).

**Comment by priseborough** [comment]
https://github.com/PX4/Firmware/pull/8298 which was submitted to test this has passed all CI tests, so this should be good to go.

**Pull request merged**